File: LC912_Sort an Array_Quick Sort.cpp
class Solution {
    int partition(vector<int>& nums, int left, int right) { // 快速排序 C++
        int pivot = nums[right]; // 选一个数值作为pivot
        int i = left - 1;
        for (int j = left; j <= right - 1; j++) { // 将数组中的数值按照选中的pivot大小进行划分，小于等于的在一起，大于的在一起
            if (nums[j] <= pivot) {
                i = i + 1;
                swap(nums[i], nums[j]); // 交换位置
            }
        }
        swap(nums[i + 1], nums[right]); // 这一行代码的意义是：把一开始就搁置的pivot再拉回来，放回中间的位置。左边是小于等于的元素，右边是大于的。
        return i + 1; // 输出pivot位置
    }
    void quicksort(vector<int>& nums, int left, int right) {
        if (left < right) { // 快速排序
            int pos = partition(nums, left, right);
            quicksort(nums, left, pos - 1);
            quicksort(nums, pos + 1, right);
        }
    }
public:
    vector<int> sortArray(vector<int>& nums) {
        quicksort(nums, 0, (int)nums.size() - 1);
        return nums;
    }
};

File: LC912_Sort an Array_Quick Sort.py
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]: # 快速排序 模板题
        # Baseline  
        if len(nums) < 2:
            return nums
        else:
            pivot = nums[0]
            greater = [i for i in nums[1:] if i > pivot]
            less = [i for i in nums[1:] if i < pivot]
            equal = [i for i in nums[1:] if i == pivot]
            return self.sortArray(less) + equal + [pivot] + self.sortArray(greater)